plugin/core/sessions.py

```diff
@@ -186,18 +186,14 @@ def on_post_exit_async(self, session: 'Session', exit_code: int, exception: Opti
         pass
 
 
-def _sequence_to_lsp_indices(sequence: Sequence[Any]) -> List[int]:
-    return list(range(1, len(sequence) + 1))
-
-
-def _enum_like_class_to_list(c: Type[object]) -> List[str]:
+def _enum_like_class_to_list(c: Type[object]) -> List[Union[int, str]]:
```

Comment:
The return type should be `Union[List[str], List[int]]`, or am I misunderstanding the use case?

Reply:
Apparently that isn't possible, because mypy can't infer which one of the list types the return value is, when it is used later. See https://github.com/sublimelsp/LSP/runs/7490372297

Reply:
Well, I guess I can add a type hint in this case.

Reply:
You can run mypy locally by running `tox` by the way

Reply:
Hm, seems that Pyright and mypy disagree what to do with the type hint comment in https://github.com/sublimelsp/LSP/pull/2000/commits/f95c500127fc5580cf9128082cce9a782371973b.

So I guess I could revert to the initial return type `List[Union[int, str]]`, or use `# type: ignore` for the failing line. Or do you see a better solution?

Thanks I will try to figure out how to run mypy locally.

Reply:
Well, in general functions that return a Union aren’t that nice. I think we can satisfy both mypy and pyright by doing a `isinstance` check but that is clumsy. We already know what we’ll pass into the function, so it seems that we need another function that basically does the same thing but returns a List[str] (the other one returning a List[int]). I think the underlying issue is that we don’t differentiate between “string enums” and “int enums”.

Reply:
We should help the type system by being explicit about what the types of those class dicts are. Below I've done some initial work for the `CompletionItemKind`. If you agree on that solution then other classes could be changed the same way.

```diff
diff --git a/plugin/core/protocol.py b/plugin/core/protocol.py
index ce2f0c7..1da857b 100644
--- a/plugin/core/protocol.py
+++ b/plugin/core/protocol.py
@@ -24,6 +24,10 @@ class DiagnosticTag:
 
 
 class CompletionItemKind:
+    @staticmethod
+    def items() -> Dict[str, int]:
+        return CompletionItemKind.__dict__
+
     Text = 1
     Method = 2
     Function = 3
diff --git a/plugin/core/sessions.py b/plugin/core/sessions.py
index 47d407e..865dfe6 100644
--- a/plugin/core/sessions.py
+++ b/plugin/core/sessions.py
@@ -53,7 +53,7 @@ from .types import DocumentSelector
 from .types import method_to_capability
 from .types import SettingsRegistration
 from .types import sublime_pattern_to_glob
-from .typing import Callable, cast, Dict, Any, Optional, List, Tuple, Generator, Iterable, Type, Protocol, Sequence, Mapping, Union  # noqa: E501
+from .typing import Callable, cast, Dict, Any, Optional, List, Tuple, Generator, Iterable, Type, Protocol, Sequence, TypeVar, Mapping, Union  # noqa: E501
 from .url import filename_to_uri
 from .url import parse_uri
 from .version import __version__
@@ -186,23 +186,26 @@ class Manager(metaclass=ABCMeta):
         pass
 
 
-def _enum_like_class_to_list(c: Type[object]) -> Union[List[int], List[str]]:
-    return [v for k, v in c.__dict__.items() if not k.startswith('_')]
+T = TypeVar("T")
+
+
+def dict_to_list(d: Dict[str, T]) -> List[T]:
+    return [v for k, v in d.items() if not k.startswith('_')]
 
 
 def get_initialize_params(variables: Dict[str, str], workspace_folders: List[WorkspaceFolder],
                           config: ClientConfig) -> dict:
-    completion_kinds = _enum_like_class_to_list(CompletionItemKind)
-    symbol_kinds = _enum_like_class_to_list(SymbolKind)
-    diagnostic_tag_value_set = _enum_like_class_to_list(DiagnosticTag)
-    completion_tag_value_set = _enum_like_class_to_list(CompletionItemTag)
-    symbol_tag_value_set = _enum_like_class_to_list(SymbolTag)
-    semantic_token_types = _enum_like_class_to_list(SemanticTokenTypes)  # type: List[str]
+    completion_kinds = dict_to_list(CompletionItemKind.items())
+    symbol_kinds = dict_to_list(SymbolKind)
+    diagnostic_tag_value_set = dict_to_list(DiagnosticTag)
+    completion_tag_value_set = dict_to_list(CompletionItemTag)
+    symbol_tag_value_set = dict_to_list(SymbolTag)
+    semantic_token_types = dict_to_list(SemanticTokenTypes)
     if config.semantic_tokens is not None:
         for token_type in config.semantic_tokens.keys():
             if token_type not in semantic_token_types:
                 semantic_token_types.append(token_type)
-    semantic_token_modifiers = _enum_like_class_to_list(SemanticTokenModifiers)
+    semantic_token_modifiers = dict_to_list(SemanticTokenModifiers)
     first_folder = workspace_folders[0] if workspace_folders else None
     general_capabilities = {
         # https://microsoft.github.io/language-server-protocol/specification#regExp
```

Reply:
`dict_to_list` -> `dict_values_to_list` would be more precise.

Reply:
But I wonder if it is reasonable to duplicate this function with the same function body just to satisfy mypy and have slightly more precise return type annotations. This function is used only a handful of times at this spot for the initialize options. I would suggest to revert to `List[Union[int, str]]`, okay?

Reply:
And to DRY it more:

```diff
diff --git a/plugin/core/protocol.py b/plugin/core/protocol.py
index ce2f0c7..58e8514 100644
--- a/plugin/core/protocol.py
+++ b/plugin/core/protocol.py
@@ -1,4 +1,4 @@
-from .typing import Any, Dict, Iterable, List, Mapping, Optional, TypedDict, Union
+from .typing import Any, Dict, Iterable, Generic, List, Mapping, Optional, TypedDict, TypeVar, Union
 from .url import filename_to_uri
 import os
 import sublime
@@ -10,6 +10,14 @@ TextDocumentSyncKindNone = 0
 TextDocumentSyncKindFull = 1
 TextDocumentSyncKindIncremental = 2
 
+T = TypeVar("T")
+
+
+class EnumLike(Generic[T]):
+    @staticmethod
+    def items() -> Dict[str, T]:
+        return CompletionItemKind.__dict__
+
 
 class DiagnosticSeverity:
     Error = 1
@@ -23,7 +31,7 @@ class DiagnosticTag:
     Deprecated = 2
 
 
-class CompletionItemKind:
+class CompletionItemKind(EnumLike[int]):
     Text = 1
     Method = 2
     Function = 3
```

Reply:
Yeah, eventually it goes into some json anyway. For now it’s okay to use List[Union] :)

Reply:
Hmm, might not work with inheritance...

Reply:
Let's postpone this to another PR and use List[Union] for now ;-)

Alternatively, there is also the enum module / [IntEnum](https://docs.python.org/release/3.4.0/library/enum.html?highlight=intenum#enum.IntEnum) which could be used via https://github.com/packagecontrol/enum dependency.

Reply:
`EnumLike(Generic[T])` looks sweet, but like jwortmann I imagined something like https://docs.python.org/3/library/enum.html. Maybe postpone that until python 3.8 because I don't like more dependencies. Or use the `EnumLike(Generic[T])` approach.

Reply:
That more concise solution using `EnumLike` didn't really work as it wouldn't be possible to access class method without instantiating an object. And `@staticmethod` doesn't have access to `cls`. So it would have to be either the more repetitive solution or some other smart approach.